src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1095,27 +1109,36 @@ void ExecuteSubmissionImpl()
                         // Create/Retrieve a configuration for each request
                         var buildRequestConfiguration = new BuildRequestConfiguration(submission.BuildRequestData, _buildParameters.DefaultToolsVersion);
                         var matchingConfiguration = _configCache.GetMatchingConfiguration(buildRequestConfiguration);
-                        var newConfiguration = ResolveConfiguration(
+                        newConfiguration = ResolveConfiguration(
                             buildRequestConfiguration,
                             matchingConfiguration,
                             submission.BuildRequestData.Flags.HasFlag(BuildRequestDataFlags.ReplaceExistingProjectInstance));
 
                         newConfiguration.ExplicitlyLoaded = true;
+                    }
 
-                        submission.BuildRequest = CreateRealBuildRequest(submission, newConfiguration.ConfigurationId);
-
+                    CacheResult cacheResult = null;
+                    // Don't lock on _syncLock to avoid calling the cache serially.
+                    // Ideally, we should lock on the <configuration, targets> tuple, but that would make the code even more convoluted
+                    // so lock just on the configuration. Realistically it should not cause overlocking because the cache is only called on
+                    // top level submissions and those tend to be unique.
+                    lock (newConfiguration)
```

Comment:
Do we actually need a lock if we successfully ensured each configuration is unique and built exactly once?